design/controlling-agent-updates.md

```diff
@@ -0,0 +1,85 @@
+# Controlling agent updates
+
+The agent is considered an extension of the service.
+As such, we have an auto-update mechanism to ensure that new features can rely on the agent support they require.
+Conceptually, it's no different from binaries on the app tier or JavaScript served to the web frontend.
+
+For instance, tasks express a minimum agent version required.
+If a selected agent doesn't meet that requirement, the agent is first asked to update itself.
+Other ways an agent can be updated include using a feature which demands an evergreen agent (YAML pipelines, for instance) or clicking the "upgrade all agents" button in a pool.
+
+For some highly-regulated customers, this kind of evergreen infrastructure is at odds with their security/compliance needs.
+- For instance, Customer "K" runs the agent inside a highly change-controlled datacenter.
+Every alteration must be approved by a high-ranking official in their organization after manual source inspection.
+Today, they diff between the release tags of their current version and the version being installed, looking for potential security issues or malicious changes.
+- Customer "N" isn't as concerned about the security aspect but is in a regulated space.
+They must satisfy their auditors that they're aware of all changes taking place inside their datacenter.
```

Comment:
Would task upgrades fall under this scope as well?  There's no model for that afaik because minor versions of tasks are consumed automatically.  I don't know if this can be fully controlled but at least logged in the build log that inputs included tasks xyz and agent version 123.

Reply:
@Joel-K We have a feature that allows you to disable in the box tasks (the ones that exist when you create an account and are auto updated) as well as marketplace tasks. In this mode only tasks installed into your account through tfx can be used by members of your org. This way you can inventory and control what tasks are installed as well as look at their min agent versions and not update until your Agent can support them. Would that work?

Reply:
If my goal was to control every aspect of a service deployed in my datacenter then the ability to whitelist only certain tasks would be very valuable.  Does this apply to specific versions of tasks or just the tasks themselves?

Personally I don't want to have to verify / approve individual versions.  That sounds like a ton of work.  I just think only pinning agent version is a half measure if you can't also pin tasks to specific minor versions.  If the task lockdown feature accomplishes that then these two features combined should help ensure all changes to the datacenter are performed explicitly.

Reply:
Thanks @Joel-K. The key message here is: we _expect_ the agent to update on our demand. That's the model for 95% (maybe 99%) of our customer base and will remain the default. For customers who - for whatever reason - can't accept that, we need a better failure mode than knocking agents offline permanently. That's what happens today.

If you use a new feature or if a task you depend on suddenly demands a new agent, those pipelines will start failing. Other pipelines should continue working (and might even have less contention for available agents :grin:). I'm arguing that's better than today, where agents start getting picked off and aren't available for any pipeline's use. I have explicit confirmation from at least one of the pseudonymous customers listed here that this is true - they're fast at getting necessary approvals, but not quite fast enough to avoid problems under today's model.